kvpb: extract requests and errors from roachpb into new package

yuzefovich opened the pull request:
This commit extracts a new `pkg/kv/kvpb` package out of `pkg/roachpb`.
We've wanted to do this for a while, and what prompted me to do this was
the desire to make the ScanResponse depend on `pkg/col/coldata`.
`api.proto` and `errors.proto` files are moved into the new package
while preserving the `roachpb` as their protobuf package. This required
also the movement of the following:
- `roachpb/gen` and `roachpb/roachpbmock` subpackages
- almost all of `api.go` (except for `roachpb.UserPriority` that was
moved into `roachpb/data.go`)
- miscellaneous error-specific files
- extracting a couple of functions from `data.go`
- `method.go`.

In order to keep the backwards-compatibility for errors that are sent
over the wire in the mixed-version clusters, for all error types that
were moved from `roachpb` to `kv/kvpb` a type migration has been
registered with the `errors` library.

Clearly, essentially all of these changes were done automatically. The
only manual change worth mentioning was when moving `errors.go`. Namely,
type conversion of `ReadWithinUncertaintyIntervalError.ObservedTimestamps`
to `roachpb.observedTimestampSlice` was removed in the loop since it
seems to not do anything.

After running `./dev gen bazel` the proto compiler in
`roachpb/BUILD.bazel` changed. I'm assuming that it is expected.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97221)
<!-- Reviewable:end -->

yuzefovich committed:
kvpb: extract requests and errors from roachpb into new package

This commit extracts a new `pkg/kv/kvpb` package out of `pkg/roachpb`.
We've wanted to do this for a while, and what prompted me to do this was
the desire to make the ScanResponse depend on `pkg/col/coldata`.
`api.proto` and `errors.proto` files are moved into the new package
while preserving the `roachpb` as their protobuf package. This required
also the movement of the following:
- `roachpb/gen` and `roachpb/roachpbmock` subpackages
- almost all of `api.go` (except for `roachpb.UserPriority` that was
moved into `roachpb/data.go`)
- miscellaneous error-specific files
- extracting a couple of functions from `data.go`
- `method.go`.

In order to keep the backwards-compatibility for errors that are sent
over the wire in the mixed-version clusters, for all error types that
were moved from `roachpb` to `kv/kvpb` a type migration has been
registered with the `errors` library.

Clearly, essentially all of these changes were done automatically. The
only manual change worth mentioning was when moving `errors.go`. Namely,
type conversion of `ReadWithinUncertaintyIntervalError.ObservedTimestamps`
to `roachpb.observedTimestampSlice` was removed in the loop since it
seems to not do anything.

After running `./dev gen bazel` the proto compiler in
`roachpb/BUILD.bazel` changed. I'm assuming that it is expected.

Epic: None

Release note: None

yuzefovich commented:
@dt thanks for volunteering! I think this is RFAL. The only unresolved question in my mind is what to do about the type migration of errors. I did add `errors.RegisterTypeMigration` as Raphael suggested, but then there were multiple failures in the [CI](https://teamcity.cockroachdb.com/viewLog.html?buildId=8734679&buildTypeId=Cockroach_BazelEssentialCi), so for now I commented it out in `kvpb/errors.go`.

knz commented on pkg/kv/kvpb/errors.go:
```diff
@@ -305,6 +306,58 @@ const (
 )
 
+// TODO: figure out whether we need this type migration, and if we do, how to
+// handle the CI failures.
```
The migrations here look correct. What are the CI failures?

yuzefovich replied:
There were a handful https://teamcity.cockroachdb.com/viewLog.html?buildId=8747339&buildTypeId=Cockroach_BazelEssentialCi. But I think it was only a couple of errors for which we defined custom wrapper encoders / decoders, and there was an incorrect order of initialization (i.e. the type migration previously would apply after the registration of encoders). I just pushed an update that should fix those errors.

yuzefovich replied:
Indeed, that was the issue, so I updated the commit. RFAL.

knz commented on pkg/kv/kvpb/errors.go:
```diff
@@ -305,6 +307,50 @@ const (
+// and are now in the kv/kvpb package.
+func init() {
+	roachpbPath := reflect.TypeOf(roachpb.Key("")).PkgPath()
+	// TODO: this doesn't have the corresponding ErrorDetailType.
```
What is this comment supposed to mean?

yuzefovich replied:
The comment on `ErrorDetailInterface` interface seems to imply that all protobuf-backed errors need to implement it, and currently `UnhandledRetryableError` and `UnhandledRetryableError` don't. Not sure how important this fact is, but I don't think it changes anything about this PR, just something to think about. I removed the TODOs though.

knz commented on pkg/kv/kvpb/errors.go:
```diff
@@ -305,6 +307,50 @@ const (
+	roachpbPath := reflect.TypeOf(roachpb.Key("")).PkgPath()
+	// TODO: this doesn't have the corresponding ErrorDetailType.
+	errors.RegisterTypeMigration(roachpbPath, "*roachpb.UnhandledRetryableError", &UnhandledRetryableError{})
+	// TODO: not sure whether internalError needs a migration.
```
it does

yuzefovich replied:
Ack, thanks.

knz reviewed (commented):
Error migration code LGTM. I did not check whether all types in the package had a corresponding registertypemigration call.

dt commented:
Looks great, thanks so much for doing this!

I am mostly just reviewing this just by the description of the automatic steps which were taken and then glancing at a few of the outputs of those steps to ensure they look right, rather than reading all 700 changed files. What I did look at is the places where it is indicated manual steps were required, though since @knz is the errors expert and already LGTM'ed that, I've left that to him.

On two of the questions raised in the PR message:

> observedTimestampSlice ...  seems to not do anything.

Indeed, it was being cast back when we passed the whole _and cast_ list to a "%v" so that the type's printer would be used, but then we iterated the list here to print a better error ourselves and the cast is indeed no longer doing anything.

> After running ./dev gen bazel the proto compiler in
> roachpb/BUILD.bazel changed. I'm assuming that it is expected.

Yep, it is: there is no longer a grpc service in pkg/roachpb so no longer need the gRPC service compiler. 

Thanks again for doing this!

yuzefovich commented:
TFTRs!

bors r+
